lib/core/versions/latest/TransactionProcessor.ts

```diff
@@ -91,60 +91,112 @@ export default class TransactionProcessor implements ITransactionProcessor {
+      let didUniqueSuffixesInMapFile: string[] = [];
+      if (mapFileModel.updateOperations !== undefined) {
+        didUniqueSuffixesInMapFile = mapFileModel.updateOperations.map(operation => operation.didUniqueSuffix);
 
-      if (!ArrayMethods.areMutuallyExclusive(anchorFile.didUniqueSuffixes, didUniqueSuffixesInMapFile)) {
-        throw new SidetreeError(ErrorCode.TransactionProcessorOperationForTheSameDidInBothAnchorAndMapFile);
+        if (!ArrayMethods.areMutuallyExclusive(anchorFile.didUniqueSuffixes, didUniqueSuffixesInMapFile)) {
+          return undefined;
+        }
       }
-    }
 
-    return mapFileModel;
+      return mapFileModel;
+    } catch (error) {
+      if (error instanceof SidetreeError) {
+        // If error is related to CAS network issues, we will surface them so retry can happen.
+        if (error.code === ErrorCode.CasNotReachable ||
+            error.code === ErrorCode.CasFileNotFound) {
+          throw error;
+        }
+
+        return undefined;
+      } else {
+        console.error(`Unexpected error fetching map file ${anchorFile.model.mapFileHash}, MUST investigate and fix: ${SidetreeError.stringify(error)}`);
+        return undefined;
+      }
+    }
   }
 
+  /**
+   * NOTE: In order to be forward-compatable with data-pruning feature,
+   * we must continue to process the operations declared in the anchor file even if the map/batch file is invalid.
+   * This means that this method MUST ONLY throw errors that are retryable (e.g. network or file not found errors),
+   * It is a design choice to hide the complexity of batch file downloading and construction within this method,
+   * instead of throwing errors and letting the caller handle them.
+   * @returns `BatchFileModel` if downloaded file is valid; `undefined` otherwise.
+   * @throws SidetreeErrors that are retryable.
+   */
   private async downloadAndVerifyBatchFile (
-    mapFile: MapFileModel
-  ): Promise<BatchFileModel> {
-    const batchFileHash = mapFile.batchFileHash;
-    console.info(`Downloading batch file '${batchFileHash}', max size limit ${ProtocolParameters.maxBatchFileSizeInBytes}...`);
+    mapFile: MapFileModel | undefined
+  ): Promise<BatchFileModel | undefined> {
+    // Can't download batch file if map file is not given.
+    if (mapFile === undefined) {
+      return undefined;
+    }
 
-    const fileBuffer = await this.downloadFileFromCas(batchFileHash, ProtocolParameters.maxBatchFileSizeInBytes);
-    const batchFileModel = await BatchFile.parse(fileBuffer);
+    try {
+      const batchFileHash = mapFile.batchFileHash;
+      console.info(`Downloading batch file '${batchFileHash}', max size limit ${ProtocolParameters.maxBatchFileSizeInBytes}...`);
+
+      const fileBuffer = await this.downloadFileFromCas(batchFileHash, ProtocolParameters.maxBatchFileSizeInBytes);
+      const batchFileModel = await BatchFile.parse(fileBuffer);
+
+      return batchFileModel;
+    } catch (error) {
```

Comment:
Nit: 3 is a pattern :) I'm not sure what is a good way to avoid the code duplication (or whether we shoud do it at all). Maybe have a function `isErrorRetryable` is enough? Up to you whether you want to fix or not.

Reply:
The two new try-catch blocks differs from the parent one in that it treats unknown errors as not retryable. So I am thinking to not move the two methods into a method.

Reply:
:thumb